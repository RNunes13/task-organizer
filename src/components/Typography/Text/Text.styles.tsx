import styled from "styled-components";
import * as Typographies from "@/theme/mixins/Typography";
import type { TextProps, TYPOGRAPHIES } from "./Text.interfaces";
import type { TypographyType } from "@/theme/mixins/Typography";

export const TextComponent = styled.p<TextProps & { $styleAs?: TYPOGRAPHIES }>`
  margin: 0;
  font: inherit;

  ${({ $styleAs, tag }) => {
    const mixins: Record<string, TypographyType> = Typographies;

    if ($styleAs && mixins[$styleAs]) {
      return mixins[$styleAs]();
    }

    if (tag && mixins[tag]) {
      return mixins[tag];
    }

    return mixins.bodyMediumRegular;
  }}

  text-align: ${({ align }) => align};
  color: ${({ color, theme }) => (color && theme.colors[color]) || "inherit"};
`;
