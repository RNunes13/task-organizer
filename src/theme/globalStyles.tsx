import { createGlobalStyle } from "styled-components";
import { CustomFonts } from "./fonts/CustomFonts";

const GlobalStyle = createGlobalStyle`
  ${CustomFonts}

  * {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
  }

  body {
    width: 100%;
    overflow-x: hidden;
    background-color: ${({ theme }) => theme.colors.bodyBg};
    color: ${({ theme }) => theme.colors.bodyColor};
  }

  button {
    cursor: pointer;
    font: inherit;
    overflow: visible;
    margin: 0;
    padding: 0;
    border: 0;
    background: transparent;
    line-height: normal;
    user-select: text;
    -moz-user-select: text;
    touch-action: manipulation;
    -ms-touch-action: manipulation;
  }

  ul, li {
    padding-left: 0;
    margin-top: 0;
    margin-bottom: 0;
    list-style: none;
  }

  #root {
    min-height: 100vh;
    display: flex;
    flex-direction: column;
  }
`;

export default GlobalStyle;
