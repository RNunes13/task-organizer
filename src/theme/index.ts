export type ThemeModeType = "dark" | "light";

enum COLORS {
  /** Base */
  base_0 = "base_0",
  base_10 = "base_10",
  base_20 = "base_20",
  base_30 = "base_30",
  base_40 = "base_40",
  base_50 = "base_50",
  base_60 = "base_60",
  base_70 = "base_70",
  base_80 = "base_80",
  base_90 = "base_90",
  base_100 = "base_100",

  /** Background */
  bodyBg = "bodyBg",
  bodyColor = "bodyColor",

  /** Primary and its variations  */
  primary_5 = "primary_5",
  primary_10 = "primary_10",
  primary_20 = "primary_20",
  primary_30 = "primary_30",
  primary_40 = "primary_40",
  primary_50 = "primary_50",
  primary_60 = "primary_60",
  primary_70 = "primary_70",
  primary_80 = "primary_80",
  primary_90 = "primary_90",
  primary = "primary",
  primaryGradientDark = "primaryGradientDark",
  primaryGradientMedium = "primaryGradientMedium",
  primaryGradientLight = "primaryGradientLight",
}

interface IColorsProps {
  base_0: string;
  base_10: string;
  base_20: string;
  base_30: string;
  base_40: string;
  base_50: string;
  base_60: string;
  base_70: string;
  base_80: string;
  base_90: string;
  base_100: string;

  bodyBg: string;
  bodyColor: string;

  primary: string;
  primary_5: string;
  primary_10: string;
  primary_20: string;
  primary_30: string;
  primary_40: string;
  primary_50: string;
  primary_60: string;
  primary_70: string;
  primary_80: string;
  primary_90: string;

  primaryGradientDark: string;
  primaryGradientMedium: string;
  primaryGradientLight: string;
}

interface IRadiusProps {
  sm: string;
  md: string;
  lg: string;
}

interface ISpacingProps {
  quarck: string;
  nano: string;
  xxxs: string;
  xxs: string;
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
  xxl: string;
  xxxl: string;
  huge: string;
  giant: string;
}

interface IIconProps {
  xsmall: string;
  small: string;
  medium: string;
  large: string;
  xlarge: string;
  xxlarge: string;
}

interface IBreakpointsProps {
  xsmall: string;
  small: string;
  medium: string;
  large: string;
  xlarge: string;
}

interface IZindexProps {
  base: number;
  top: number;
  modal: number;
}

interface IFontsProps {
  baseFontFamily: string;
  secondaryFontFamily: string;
}

interface ITransitionsProps {
  standardEase: string;
  accelerateEase: string;
  decelerateEase: string;
}

interface IThemeProps {
  radius: IRadiusProps;
  colors: IColorsProps;
  spaces: ISpacingProps;
  icon: IIconProps;
  breakpoint: IBreakpointsProps;
  fonts: IFontsProps;
  index: IZindexProps;
  transitions: ITransitionsProps;
}

export { theme, COLORS };

export interface IMixinProps {
  theme: IThemeProps;
}

export type {
  IThemeProps,
  IColorsProps,
  ISpacingProps,
  IBreakpointsProps,
  IFontsProps,
};

const baseColors: IColorsProps = {
  base_100: '#17181F',
  base_90: '#2C2D34',
  base_80: '#42434B',
  base_70: '#595A63',
  base_60: '#72737B',
  base_50: '#8B8C95',
  base_40: '#A5A6B0',
  base_30: '#C1C1CB',
  base_20: '#DCDDE7',
  base_10: '#F9FAFF',
  base_0: '#FFFFFF',

  bodyBg: "#17181F",
  bodyColor: "#FFFFFF",

  primary: '#fc76a1',
  primary_5: "#fee3ec",
  primary_10: "#fed5e2",
  primary_20: "#fdc8d9",
  primary_30: "#fdacc6",
  primary_40: "#fc91b3",
  primary_50: "#fc76a1",
  primary_60: "#e26a90",
  primary_70: "#c95e80",
  primary_80: "#974660",
  primary_90: "#7e3b50",

  primaryGradientDark: '#ce69e0',
  primaryGradientMedium: '#f669aa',
  primaryGradientLight: '#fe9483',
};

const lightColors: Partial<IColorsProps> = {
  bodyBg: baseColors.base_10,
  bodyColor: baseColors.base_100,
};

const darkColors: Partial<IColorsProps> = {
  bodyBg: baseColors.base_100,
  bodyColor: baseColors.base_10,
};

const theme: IThemeProps = {
  colors: baseColors,
  radius: {
    sm: '8px',
    md: '12px',
    lg: '20px',
  },
  index: {
    base: 0,
    top: 10,
    modal: 100,
  },
  spaces: {
    quarck: '4px',
    nano: '8px',
    xxxs: '12px',
    xxs: '16px',
    xs: '20px',
    sm: '24px',
    md: '32px',
    lg: '40px',
    xl: '48px',
    xxl: '56px',
    xxxl: '64px',
    huge: '80px',
    giant: '96px',
  },
  breakpoint: {
    xsmall: "576px",
    small: "768px",
    medium: "992px",
    large: "1200px",
    xlarge: "1350px",
  },
  icon: {
    xsmall: "12px",
    small: "16px",
    medium: "24px",
    large: "32px",
    xlarge: "64px",
    xxlarge: "96px",
  },
  fonts: {
    secondaryFontFamily: "HirukoProBook",
    baseFontFamily:
      'Montserrat, "Helvetica Neue", "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", BlinkMacSystemFont, -apple-system, sans-serif',
  },
  transitions: {
    standardEase: 'cubic-bezier(0.4, 0, 0.2, 1)',
    accelerateEase: 'cubic-bezier(0.4, 0, 1, 1)',
    decelerateEase: 'cubic-bezier(0, 0, 0.2, 1)',
  }
};

export function getTheme(themeMode: ThemeModeType): IThemeProps {
  const modeColors = themeMode === "dark" ? darkColors : lightColors;

  return {
    ...theme,
    colors: {
      ...theme.colors,
      ...modeColors,
    },
  };
}

export default theme;
