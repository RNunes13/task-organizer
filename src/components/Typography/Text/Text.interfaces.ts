import { COLORS } from "@/theme";
import * as typographies from "@/theme/mixins/Typography";

export type ALIGNMENTS  = "left" | "center" | "right";

export type TYPOGRAPHIES = keyof typeof typographies;

export interface TextProps {
  align?: ALIGNMENTS;
  as?: any;
  children?: any;
  className?: string;
  color?: COLORS;
  tag?: string;
  styleAs?: TYPOGRAPHIES;
}
