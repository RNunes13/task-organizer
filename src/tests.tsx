import userEvent from "@testing-library/user-event";
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { AppContext } from "@/app/App.provider";
import { theme } from "@/theme";
import type { AppContextProps } from "@/app/App.interfaces";

interface ProvidersProps {
  children: React.ReactElement;
  appContext?: Partial<AppContextProps>;
}

interface RenderProviderProps {
  wrapperProps?: Omit<ProvidersProps, "children">;
}

const defaultAppContext: AppContextProps = {
  searchTerm: null,
  themeMode: "dark",
  dispatch: jest.fn(),
};

const Providers: React.FC<ProvidersProps> = ({
  children,
  appContext = defaultAppContext,
}) => (
  <AppContext.Provider value={{ ...defaultAppContext, ...appContext }}>
    <ThemeProvider theme={theme}>
      <BrowserRouter>{children}</BrowserRouter>
    </ThemeProvider>
  </AppContext.Provider>
);

export async function actWait(amount = 0) {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, amount));
  });
}

const customRender = (
  ui: React.ReactElement,
  options?: RenderOptions & RenderProviderProps
) => {
  const { wrapperProps = {}, ...rest } = options || {};

  return {
    user: userEvent.setup(),
    ...render(ui, {
      wrapper: ({ children }) => (
        <Providers {...wrapperProps}>{children}</Providers>
      ),
      ...rest,
    }),
  };
};

export const renderWithTheme = (component: any): ReactTestRenderer => (
  renderer.create(<Providers>{component}</Providers>)
);

export * from "@testing-library/react";
export { customRender as render };
