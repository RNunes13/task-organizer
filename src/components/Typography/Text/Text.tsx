import { FC } from 'react';
import { TextComponent } from './Text.styles';
import { TextProps } from './Text.interfaces';

export const Text: FC<TextProps> = ({ as, tag, styleAs, ...rest }) => (
  <TextComponent {...rest} as={as || tag} $styleAs={styleAs} />
);

Text.defaultProps = {
  tag: 'p'
};
