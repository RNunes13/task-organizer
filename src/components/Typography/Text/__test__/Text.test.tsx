import { COLORS } from "@/theme";
import { renderWithTheme } from "@/tests";
import { Text } from "../Text";
import { TextProps } from "../Text.interfaces";

const defaulProps: TextProps = {
  children: "Text",
};

const getComponent = (props: Partial<TextProps> = {}) =>
  renderWithTheme(<Text {...defaulProps} {...props} />).toJSON();

describe("Text component", () => {
  it("renders correctly", () => {
    const tree = getComponent({
      tag: "p",
      styleAs: "bodyMediumRegular",
      color: COLORS.primary,
    });
    expect(tree).toMatchSnapshot();
  });

  it("renders heading 1", () => {
    const tree = getComponent({
      tag: "h1",
    });
    expect(tree).toMatchSnapshot();
  });

  it("renders heading 2 styled as heading 3", () => {
    const tree = getComponent({
      tag: "h2",
      styleAs: "h3",
    });
    expect(tree).toMatchSnapshot();
  });

  it("renders heading 3, styled as heading 4, and primaryGradientMedium color", () => {
    const tree = getComponent({
      tag: "h3",
      styleAs: "h4",
      color: COLORS.primaryGradientMedium,
    });
    expect(tree).toMatchSnapshot();
  });

  it("renders heading 4, styled as heading 5, primary color, and aligned to the right", () => {
    const tree = getComponent({
      tag: "h4",
      styleAs: "h5",
      color: COLORS.primary,
      align: "right",
    });
    expect(tree).toMatchSnapshot();
  });
});
