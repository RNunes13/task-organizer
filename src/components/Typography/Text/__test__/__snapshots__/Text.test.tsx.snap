// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component renders correctly 1`] = `
.c0 {
  margin: 0;
  font: inherit;
  font-family: Montserrat,"Helvetica Neue","Segoe UI",Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  line-height: 24px;
  color: #fc76a1;
}

<p
  className="c0"
  color="primary"
>
  Text
</p>
`;

exports[`Text component renders heading 1 1`] = `
.c0 {
  margin: 0;
  font: inherit;
  font-family: Montserrat,"Helvetica Neue","Segoe UI",Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  line-height: 24px;
  color: inherit;
}

<h1
  className="c0"
>
  Text
</h1>
`;

exports[`Text component renders heading 2 styled as heading 3 1`] = `
.c0 {
  margin: 0;
  font: inherit;
  font-family: HirukoProBook;
  font-style: normal;
  font-weight: 400;
  font-size: 20px;
  line-height: 28px;
  color: inherit;
}

<h2
  className="c0"
>
  Text
</h2>
`;

exports[`Text component renders heading 3, styled as heading 4, and primaryGradientMedium color 1`] = `
.c0 {
  margin: 0;
  font: inherit;
  font-family: Montserrat,"Helvetica Neue","Segoe UI",Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 700;
  font-size: 16px;
  line-height: 24px;
  color: #f669aa;
}

<h3
  className="c0"
  color="primaryGradientMedium"
>
  Text
</h3>
`;

exports[`Text component renders heading 4, styled as heading 5, primary color, and aligned to the right 1`] = `
.c0 {
  margin: 0;
  font: inherit;
  font-family: Montserrat,"Helvetica Neue","Segoe UI",Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 700;
  font-size: 14px;
  line-height: 20px;
  text-align: right;
  color: #fc76a1;
}

<h4
  align="right"
  className="c0"
  color="primary"
>
  Text
</h4>
`;
